### DISCUSSION TITLE
Can you reverse WSL GUI functionality - translating Windows code to Linux code?

### DISCUSSION THREAD

**Discussion by ghost** [microsoft/WSL#6153] [General] [answered]
It seems like running Linux GUI application is coming soon in Microsoft's WSL.

I'm a complete novice so please excuse me for my ignorance, first. Since there seems to be some form of translation to allow Linux GUI apps to run in Windows, I'm curious if we could translate in "reverse": run Windows app inside WSL. Granted WSL is exposed only through command line, but perhaps it can be extended to Linux proper. I'm picturing a basic lookup of table for converting Linux to Windows code, but I'm sure it's much more complex than this.

I know that there's Wine for this, but it would be nice if Microsoft officially supported this feature.

**Comment by ghost** [comment, marked as answer]
Why would you want to run Windows Apps in WSL, when you have Windows?

If you meant to ask "Can I interact with the WSL filesystem with Windows Apps in bash?":

you can run:

```bash
$ notepad.exe filename.txt
```

to edit "filename.txt" which is in the Linux filesystem using notepad

**Reply by handicraftsman** [reply]
While running it in WSL is definitely a weird idea, there would be sense for running Windows apps in bare metal Linux environments, due to:
1. Linux being able to run on low-end or just old machines easier
2. Linux being the main OS for some people, either due to personal choices, software or performance requirements. For example, i find Linux apps more useful and intuitive for me (i am into cli based pc usage way), and i also cant run Windows all the time my PC is running due to having only 4 GB RAM and yet having to run RAM heavy tasks, such as Webpack (yes! Its heavy on such machines!).

Please note that it is definitely doable, although it requires writing kernel modules on both Linux and NT kernel sides.

Here i mention WSL2 though, with its VM approach.

The solution i think about:
Write kernel modules for NT and Linux which would allow sharing userspace memory, filesystems, and UNIX sockets (essential for X11 and Wayland) from and to linux in a performant way, then plug this into an existing virtual machine implementation. This would also allow using unmodified Windows ISO files to setup the reverse-WSL.
Userspace apps would use Linux alloc, fs, and unix sockets, while NT kernel, for its internal use, would use pure VM memory.

For the end user this will look as:
1. Insmod the linux side module
2. Create a VM instance
3. Make it install Windows OS in pure virtual mode, without resource bridging
4. VM runs an own app (possibly inside of itself) which installs the NT driver to the Windows setup.
5. VM boots into Windows, with possible bridging of Windows app windows to X11 or Wayland, similar to how Parallels do that on macOS.
